### PR TITLE
Consistent snapshots

### LINKED ISSUE #8: Consistent snapshots
The CLI should default to using consistent snapshots.

### PR DESCRIPTION
**Pull request by lmars**
The idea here is that repositories will default to using consistent snapshots, and can be disabled by running `tuf init --consistent-snapshot=false` before adding any target files.

If target files have already been added, the only way to change whether consistent snapshots are used is to run (the yet unimplemented) `tuf regenerate --consistent-snapshot=false`.

Closes #8.

### CODE REVIEW

File: client/remote_store.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+	if res.StatusCode == http.StatusNotFound {
+		res.Body.Close()
+		return nil, 0, ErrNotFound{path}
+	} else if res.StatusCode != http.StatusOK {
+		res.Body.Close()
+		return nil, 0, fmt.Errorf("unexpected HTTP response code: %d", res.StatusCode)
+	}
+
+	size, err := strconv.ParseInt(res.Header.Get("Content-Length"), 10, 0)
```
Comment (titanous) [inline client/remote_store.go]: We can't assume that `Content-Length` will be present, since the response might be chunked.

File: repo.go [hunk 2]
```diff
@@ -470,6 +484,34 @@ func (r *Repo) TimestampWithExpires(expires time.Time) error {
...
+	addHashes := func(name string, meta data.Files) {
+		if m, ok := meta[name]; ok {
+			hashes[name] = m.Hashes
+		}
+	}
+	timestamp, err := r.timestamp()
+	if err != nil {
+		return nil, err
+	}
+	snapshot, err := r.snapshot()
```
Comment (titanous) [inline repo.go]: Now that we're doing multiple lookups, does it make sense to memoize these accessors?

File: tuf/init.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package main
+
+import (
+	"github.com/flynn/go-docopt"
+	"github.com/flynn/go-tuf"
+)
+
+func init() {
+	register("init", cmdInit, `
+usage: tuf init [--consistent-snapshot=false]
```
Comment (titanous) [inline tuf/init.go]: Should this default to `true`? Despite the docs I expect people will call it accidentally/redundantly.

**Comment by titanous** [comment]
Are we doing interop tests for both consistent hashing and not?

File: client/remote_store.go (hunk 1 above)
Comment (lmars) [inline client/remote_store.go]: Thats why I return `-1` below to indicate that fact, or is that logic flawed?

File: repo.go (hunk 2 above)
Comment (lmars) [inline repo.go]: So the issue here is the memoized versions getting out of date. For example if data gets signed or thresholds change, the correct invalidation needs to happen. It is possible but I find it less complex to just decode these each time. WDYT?

File: client/remote_store.go (hunk 1 above)
Comment (titanous) [inline client/remote_store.go]: Looks right, must have missed that the first time.

File: tuf/init.go (hunk 3 above)
Comment (lmars) [inline tuf/init.go]: The default will be true, the only way it will be false is if the flag is explicitly set to `false`

File: repo.go (hunk 2 above)
Comment (titanous) [inline repo.go]: Okay, that's fine.

**Comment by lmars** [comment]
@titanous interop tests for consistent snapshots added

File: client/remote_store.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if res.StatusCode == http.StatusNotFound {
+		res.Body.Close()
+		return nil, 0, ErrNotFound{s}
+	} else if res.StatusCode != http.StatusOK {
+		res.Body.Close()
+		return nil, 0, fmt.Errorf("unexpected HTTP response code: %d", res.StatusCode)
```
Comment (titanous) [inline client/remote_store.go]: Return a [`url.Error`](https://golang.org/pkg/net/url/#Error) here.

**Comment by titanous** [comment]
LGTM pending my last comment.

**Pull request merged**